numba/npyufunc/_internal.c

```diff
@@ -104,6 +105,548 @@ PyTypeObject PyUFuncCleaner_Type = {
+
+static void
+dufunc_dealloc(PyDUFuncObject *self)
+{
+    /* Note: There is no need to call PyArray_free() on
+       self->ufunc->ptr, since ufunc_dealloc() will do it for us. */
+    Py_XDECREF(self->ufunc);
+    Py_XDECREF(self->dispatcher);
+    Py_XDECREF(self->keepalive);
+    Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static PyObject *
+dufunc_repr(PyDUFuncObject *dufunc)
+{
+    return PyString_FromFormat("<numba._DUFunc '%s'>", dufunc->ufunc->name);
+}
+
+static PyObject *
+dufunc_call(PyDUFuncObject *self, PyObject *args, PyObject *kws)
+{
+    PyObject *result=NULL;
+
+    result = PyUFunc_Type.tp_call((PyObject *)self->ufunc, args, kws);
+    if ((!self->frozen) &&
+            (result == NULL) &&
+            (PyErr_Occurred()) &&
+            (PyErr_ExceptionMatches(PyExc_TypeError))) {
+
+        /* Break back into Python when we fail at dispatch. */
+        PyErr_Clear();
+        PyObject *method = PyObject_GetAttrString(
+            (PyObject*)self, "_compile_for_args");
+
+        if (method) {
+            result = PyObject_Call(method, args, kws);
+            if (result) {
+                Py_DECREF(result);
+                result = PyUFunc_Type.tp_call((PyObject *)self->ufunc, args,
+                                              kws);
+            }
+        }
+        Py_XDECREF(method);
+    }
+    return result;
+}
+
+static Py_ssize_t
+_get_nin(PyObject * py_func_obj)
+{
+    int result = -1;
+    PyObject *inspect=NULL, *getargspec=NULL, *argspec=NULL, *args=NULL;
+
+    inspect = PyImport_ImportModule("inspect");
+    if (!inspect) goto _get_nin_cleanup;
+    getargspec = PyObject_GetAttrString(inspect, "getargspec");
+    if (!getargspec) goto _get_nin_cleanup;
+    argspec = PyObject_CallFunctionObjArgs(getargspec, py_func_obj, NULL);
+    if (!argspec) goto _get_nin_cleanup;
+    args = PyObject_GetAttrString(argspec, "args");
+    if (!args) goto _get_nin_cleanup;
+    result = PyList_Size(args);
+
+  _get_nin_cleanup:
+    Py_XDECREF(args);
+    Py_XDECREF(argspec);
+    Py_XDECREF(getargspec);
+    Py_XDECREF(inspect);
+    return result;
+}
+
+static int
+dufunc_init(PyDUFuncObject *self, PyObject *args, PyObject *kws)
+{
+    PyObject *dispatcher=NULL, *keepalive=NULL, *py_func_obj=NULL, *tmp;
+    PyUFuncObject *ufunc=NULL;
+    int identity=PyUFunc_None;
+    int nin=-1, nout=1;
+    char *name=NULL, *doc=NULL;
+
+    static char * kwlist[] = {"dispatcher", "identity", "keepalive", "nin",
+                              "nout", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kws, "O|iO!nn", kwlist,
+                                     &dispatcher, &identity,
+                                     &PyList_Type, &keepalive, &nin, &nout)) {
+        return -1;
+    }
+
+    py_func_obj = PyObject_GetAttrString(dispatcher, "py_func");
+    if (!py_func_obj) {
+        return -1;
+    }
+
+    if (nin < 0) {
+        nin = (int)_get_nin(py_func_obj);
+        if ((nin < 0) || (PyErr_Occurred())) {
```

Comment:
Again, this is defensive programming.  Even though I wrote this, it does mean some one will come along and make a change later that breaks this.